lib/ts/utils/dateProvider/index.ts

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2024, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import WindowHandlerReference from "../windowHandler";
+import { DateProvider, defaultDateProviderImplementation } from "./defaultImplementation";
+import { DateProviderInterface, DateProviderInput } from "./types";
+
+/**
+ * A utility class for managing a reference to a DateProviderInterface, allowing customization of time-related operations.
+ * This class is designed to address clock skew issues between the server and client by providing a consistent mechanism
+ * for obtaining current time adjusted for clock skew.
+ *
+ * @class DateProviderReference
+ */
+export default class DateProviderReference {
+    private static instance?: DateProviderReference;
+
+    dateProvider: DateProviderInterface;
+
+    constructor(dateProviderInput?: DateProviderInput) {
+        let dateProviderFunc: DateProviderInput = original => original;
+        if (dateProviderInput !== undefined) {
+            dateProviderFunc = dateProviderInput;
+            this.dateProvider = dateProviderFunc(defaultDateProviderImplementation);
+        } else {
+            this.dateProvider = dateProviderFunc(defaultDateProviderImplementation);
+            // Initialise the clockSkewInMillis from localStorage for the default implementation.
+            const localStorage = WindowHandlerReference.getReferenceOrThrow().windowHandler.localStorage;
+            const clockSkewInMillis = parseInt(localStorage.getItemSync(DateProvider.CLOCK_SKEW_KEY) || "0", 10);
+            this.dateProvider.setClientClockSkewInMillis(clockSkewInMillis);
```

Comment:
I'd prefer if this was done in a factory function (a static `init`) in the default date provider impl class.
We can then call that factory function here instead of exporting an instance from there.
Please also comment that we use an `init` call to build the instance because we need it to execute after the window handler is initialized.

Reply:
I have made the changes but please do check if that is what you were trying to suggest.